File: android/src/withMaps/java/org/ligi/passandroid/LocationsMapFragment.java
package org.ligi.passandroid;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import java.util.List;
import org.ligi.axt.AXT;
import org.ligi.passandroid.model.pass.PassLocation;
import org.ligi.passandroid.ui.PassViewActivityBase;

public class LocationsMapFragment extends SupportMapFragment {

    private PassViewActivityBase base_activity;
    public boolean click_to_fullscreen = false;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View root = super.onCreateView(inflater, container, savedInstanceState);

        base_activity = (PassViewActivityBase) getActivity();

        if (!(getActivity() instanceof PassViewActivityBase)) {
            throw new IllegalArgumentException("LocationsMapFragment must be used inside a PassViewActivityBase");
        }

        getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(final GoogleMap map) {

                map.setOnMapLoadedCallback(new GoogleMap.OnMapLoadedCallback() {
                    @Override
                    public void onMapLoaded() {
                        if (click_to_fullscreen) map.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
                            @Override
                            public void onMapClick(final LatLng latLng) {
                                App.component().passStore().setCurrentPass(base_activity.currentPass);
                                AXT.at(getActivity()).startCommonIntent().activityFromClass(FullscreenMapActivity.class);
                            }
                        });


                        LatLngBounds.Builder boundBuilder = new LatLngBounds.Builder();

                        final List<PassLocation> locations = base_activity.currentPass.getLocations();

                        for (PassLocation l : locations) {

                            // yea that looks stupid but need to split LatLng free/nonfree - google play services ^^
                            final LatLng latLng = new LatLng(l.lat, l.lon);
                            map.addMarker(new MarkerOptions().position(latLng).title(l.getName(base_activity.currentPass))
                                          //.icon(BitmapDescriptorFactory.fromBitmap(base_activity.passbook.getIconBitmap())));
                            );

                            boundBuilder = boundBuilder.include(latLng);
                        }

                        map.setOnInfoWindowClickListener(new GoogleMap.OnInfoWindowClickListener() {
                            @Override
                            public void onInfoWindowClick(Marker marker) {
                                final Intent i = new Intent();
                                i.setAction(Intent.ACTION_VIEW);
                                i.setData(Uri.parse("geo:" + marker.getPosition().latitude + "," + marker.getPosition().longitude + "?q=" + marker.getTitle()));
                                getActivity().startActivity(i);
                            }
                        });
                        map.moveCamera(CameraUpdateFactory.newLatLngBounds(boundBuilder.build(), 100));

                        // limit zoom-level to 17 - otherwise we could be so zoomed in that it looks buggy
                        map.moveCamera(CameraUpdateFactory.zoomTo(Math.min(17f, map.getCameraPosition().zoom)));

                    }
                });

            }
        });

        return root;
    }
}

File: android/src/withMaps/java/org/ligi/passandroid/LocationsMapFragment.kt
package org.ligi.passandroid

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.net.toUri
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.MarkerOptions
import org.koin.android.ext.android.inject
import org.ligi.kaxt.startActivityFromClass
import org.ligi.passandroid.model.PassStore
import org.ligi.passandroid.ui.PassViewActivityBase
import kotlin.math.min

class LocationsMapFragment : SupportMapFragment() {
    var clickToFullscreen = false

    val passStore : PassStore by inject()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = super.onCreateView(inflater, container, savedInstanceState)
        val baseActivity = activity as PassViewActivityBase

        getMapAsync { map ->
            map.setOnMapLoadedCallback {
                if (clickToFullscreen)
                    map.setOnMapClickListener {
                        passStore.currentPass = baseActivity.currentPass
                        baseActivity.startActivityFromClass(FullscreenMapActivity::class.java)
                    }


                var boundBuilder = LatLngBounds.Builder()

                val locations = baseActivity.currentPass.locations

                for (l in locations) {

                    // yea that looks stupid but need to split LatLng free/nonfree - google play services ^^
                    val latLng = LatLng(l.lat, l.lon)
                    val marker = MarkerOptions().position(latLng).title(l.getNameWithFallback(baseActivity.currentPass))
                    map.addMarker(marker)


                    boundBuilder = boundBuilder.include(latLng)
                }

                map.setOnInfoWindowClickListener { marker ->
                    val i = Intent()
                    i.action = Intent.ACTION_VIEW
                    i.data = ("geo:" + marker.position.latitude + "," + marker.position.longitude + "?q=" + marker.title).toUri()
                    baseActivity.startActivity(i)
                }
                map.moveCamera(CameraUpdateFactory.newLatLngBounds(boundBuilder.build(), 100))

                // limit zoom-level to 17 - otherwise we could be so zoomed in that it looks buggy
                map.moveCamera(CameraUpdateFactory.zoomTo(min(17f, map.cameraPosition.zoom)))
            }
        }

        return root
    }
}